worker/modelcache/worker_test.go

```diff
@@ -449,6 +442,79 @@ func (s *WorkerSuite) TestRemoveUnit(c *gc.C) {
 	}
 }
 
+func (s *WorkerSuite) TestWatcherErrorCacheMarkSweep(c *gc.C) {
+	// Some state to close over.
+	fakeModelSent := false
+	errorSent := false
+
+	s.config.WatcherFactory = func() modelcache.BackingWatcher {
+		return testingMultiwatcher{
+			Multiwatcher: s.StatePool.SystemState().WatchAllModels(s.StatePool),
+			manipulate: func(deltas []multiwatcher.Delta) ([]multiwatcher.Delta, error) {
+				if !fakeModelSent || !errorSent {
+					for _, delta := range deltas {
+						// The first time we see a model, add an extra model delta.
+						// This will be cached even though it does not exist in state.
+						if delta.Entity.EntityId().Kind == "model" && !fakeModelSent {
+							fakeModelSent = true
+							return append(deltas, multiwatcher.Delta{
+								Entity: &multiwatcher.ModelInfo{
+									ModelUUID: "fake-ass-model-uuid",
+									Name:      "evict-this-cat",
+								},
+							}), nil
+						}
+
+						// The first time we see an application, throw an error.
+						// This will restart the watcher and cause a cache refresh.
+						// We expect after this that the application will reside in
+						// the cache and our fake model will be removed.
```

Comment:
what about the cascade issue? That an application may need to be removed because the model was gone. Is that addressed elsewhere? I know I saw things like 'order events by inverse uint64 id'.

Reply:
I did solve this in my example of mark, sweep, so that during the changes loop, a model removal also removed an application and other following entities... as long as the subsequent calls to removing of the model application become a noop, it shouldn't matter. The danger is cyclic events whilst cascading.

Reply:
Any time an entity is cached without its model present, a place-holder model is cached first in anticipation of it arriving presently, so it gets a lower uint64 ID.

By ensuring sweep goes in descending ID order, we always ensure that a model is removed after its children and the the resident manager needs know nothing about specific types or hierarchies.

Reply:
Because this was split into two PRs it was hard to conceptualise, but makes sense once viewed as a hole.